### PR TITLE
Add decline code constants

### PR DESCRIPTION
**Pull request by daisy1754**
Adding constants for code in https://stripe.com/docs/declines/codes

### CODE REVIEW

File: error.go [hunk 1]
```diff
@@ -115,6 +117,53 @@ const (
 	ErrorCodeInvalidSwipeData ErrorCode = "invalid_swipe_data"
 )
 
+// List of DeclineCode values.
+const (
+	DeclineCodeApprovalWithID                 DeclineCode = "approve_with_id"
```
Comment (remi-stripe) [inline error.go]: This should be DeclineCodeApproveWithID

File: error.go [hunk 2]
```diff
@@ -115,6 +117,53 @@ const (
 	ErrorCodeInvalidSwipeData ErrorCode = "invalid_swipe_data"
 )
 
+// List of DeclineCode values.
+const (
```
Comment (remi-stripe) [inline error.go]: You're missing `authentication_required`

File: error.go [hunk 3]
```diff
@@ -115,6 +117,53 @@ const (
...
+	DeclineCodeDoNotHonor                     DeclineCode = "do_not_honor"
+	DeclineCodeDoNotTryAgain                  DeclineCode = "do_not_try_again"
+	DeclineCodeDuplicateTransaction           DeclineCode = "duplicate_transaction"
+	DeclineCodeExpiredCard                    DeclineCode = "expired_card"
+	DeclineCodeFraudulent                     DeclineCode = "fraudulent"
+	DeclineCodeGenericDecline                 DeclineCode = "generic_decline"
+	DeclineCodeIncorrectNumber                DeclineCode = "incorrect_number"
+	DeclineCodeIncorrectCVC                   DeclineCode = "incorrect_cvc"
+	DeclineCodeIncorrectPIN                   DeclineCode = "incorrect_pin"
+	DeclineCodeIncorrectZIP                   DeclineCode = "incorrect_zip"
```
Comment (remi-stripe) [inline error.go]: We use `Zip` and not `ZIP` in the library

File: error.go [hunk 4]
```diff
@@ -115,6 +117,53 @@ const (
...
+	DeclineCodeCurrencyNotSupported           DeclineCode = "currency_not_supported"
+	DeclineCodeDoNotHonor                     DeclineCode = "do_not_honor"
+	DeclineCodeDoNotTryAgain                  DeclineCode = "do_not_try_again"
+	DeclineCodeDuplicateTransaction           DeclineCode = "duplicate_transaction"
+	DeclineCodeExpiredCard                    DeclineCode = "expired_card"
+	DeclineCodeFraudulent                     DeclineCode = "fraudulent"
+	DeclineCodeGenericDecline                 DeclineCode = "generic_decline"
+	DeclineCodeIncorrectNumber                DeclineCode = "incorrect_number"
+	DeclineCodeIncorrectCVC                   DeclineCode = "incorrect_cvc"
+	DeclineCodeIncorrectPIN                   DeclineCode = "incorrect_pin"
```
Comment (remi-stripe) [inline error.go]: @brandur-stripe Should we use `PIN` or `Pin`?

File: error.go [hunk 5]
```diff
@@ -115,6 +117,53 @@ const (
...
+	DeclineCodeIncorrectZIP                   DeclineCode = "incorrect_zip"
+	DeclineCodeInsufficientFunds              DeclineCode = "insufficient_funds"
+	DeclineCodeInvalidAccount                 DeclineCode = "invalid_account"
+	DeclineCodeInvalidAmount                  DeclineCode = "invalid_amount"
+	DeclineCodeInvalidCvc                     DeclineCode = "invalid_cvc"
```
Comment (remi-stripe) [inline error.go]: This should be `DeclineCodeInvalidCVC`

**Comment by remi-stripe** [comment]
@daisy1754 Thanks a lot for doing the work on that one! I left some comments but this is already looking great!

**Comment by remi-stripe** [comment]
@daisy1754 Thanks for making all the changes! Can I ask you to squash all your commits into one? That's what we usually do on the library before merging. Once it's done, you can assign to @brandur-stripe for review!

**Comment by daisy1754** [comment]
@remi-stripe  thanks for quick turnaround and detailed review. I squashed all commits into one. It seems I don't have permission to assign PR to another person

File: error.go (hunk 4 above)
Comment (remi-stripe) [reply error.go]: let's use PIN, it's an acronym

File: error.go [hunk 6]
```diff
@@ -115,6 +118,54 @@ const (
...
+	DeclineCodeInvalidNumber                  DeclineCode = "invalid_number"
+	DeclineCodeInvalidPIN                     DeclineCode = "invalid_pin"
+	DeclineCodeIssuerNotAvailable             DeclineCode = "issuer_not_available"
+	DeclineCodeLostCard                       DeclineCode = "lost_card"
+	DeclineCodeMerchantBlacklist              DeclineCode = "merchant_blacklist"
+	DeclineCodeNewAccountInformationAvailable DeclineCode = "new_account_information_available"
+	DeclineCodeNoActionTaken                  DeclineCode = "no_action_taken"
+	DeclineCodeNotPermitted                   DeclineCode = "not_permitted"
+	DeclineCodePickupCard                     DeclineCode = "pickup_card"
+	DeclineCodePinTryExceeded                 DeclineCode = "pin_try_exceeded"
```
Comment (remi-stripe) [inline error.go]: @daisy1754 Can you rename `Pin` to `PIN` to be consistent?

Comment (daisy1754) [reply error.go]: @remi-stripe  done

**Comment by remi-stripe** [comment]
@daisy1754 Thanks a lot! Re-assigning to @brandur-stripe to do a final review and then a release!

**Comment by brandur-stripe** [comment]
LGTM! Thank you @daisy1754 for the very close attention to detail here. Great stuff :)

@remi-stripe You may have noted elsewhere otherwise, but I think technically this is a breaking change. Now that `DeclineCode` on `CardError` has transition from `string` to the new `DeclineCode` type, users that were putting it in a string field may now have to convert it back to a string to do so:

``` go
// Before
log.DeclineCode = stripeErr.DeclineCode

// After
log.DeclineCode = string(stripeErr.DeclineCode)
```

Any thoughts on what to do about this? We could just release the major. Another alternative would be to not change the type of `CardError.DeclineCode` for the time being (leave it as a `string` for now), and then change it over on the next major as a separate step. It wouldn't be as convenient or type safe, but having access to the new constants would still be useful for people. Thoughts?

**Comment by remi-stripe** [comment]
Ah yes nice catch, we likely should leave as a string for now with a TODO for the next major version (there are a few other things to clean up!)

File: error.go [hunk 7]
```diff
@@ -191,7 +242,7 @@ func (e *PermissionError) Error() string {
 // reason.
 type CardError struct {
 	stripeErr   *Error
-	DeclineCode string `json:"decline_code,omitempty"`
+	DeclineCode DeclineCode `json:"decline_code,omitempty"`
```
Comment (brandur-stripe) [inline error.go]: Would you mind changing this line to something like:

``` go
	// DeclineCode is a code indicating a card issuer's reason for declining a
	// card (if they provided one).
	//
	// TODO: This field should be of type DeclineCode (as defined above), but
	// we've let it as a string for now for backwards compatibility. Change its
	// type on the library's next major version.
	DeclineCode string `json:"decline_code,omitempty"`
}
```

File: error.go [hunk 8]
```diff
@@ -232,5 +283,5 @@ type rawError struct {
 // type to help with deserialization. (e.g. DeclineCode).
 type rawErrorInternal struct {
 	*Error
-	DeclineCode *string `json:"decline_code,omitempty"`
+	DeclineCode *DeclineCode `json:"decline_code,omitempty"`
```
Comment (brandur-stripe) [inline error.go]: And similar to the above, probably this one too.

**Comment by brandur-stripe** [comment]
Cool, thanks Remi!

@daisy1754 As noted above, we're probably going to have to ship this out in two phases just so we can send the first one out as a minor version instead of having to cut a major. I left some suggested wording for a comment to add to the code above. Let me know if that all makes sense.

After that change is finished, I think this is good to go! Thanks.

File: error.go (hunk 7 above)
Comment (daisy1754) [reply error.go]: hm good catch. what do you think about using type alias (ie., `type DeclineCode = string`)

Comment (brandur-stripe) [reply error.go]: Good question. I'd still say we should just put a `string`: the long term intent is that `DeclineCode` not be a type alias, so I'd say we just make it the right thing now, and then keep it that way.

Admittedly, we have to change something either way on the next major, but omitting the type alias in favor of a simple `string` makes looking at what the field's definition more clear in my mind.

**Comment by daisy1754** [comment]
@brahn-stripe  updated

**Comment by brandur-stripe** [comment]
This looks good to me. Thank you again for the persistent effort on this one @daisy1754!

**Comment by brandur-stripe** [comment]
Released as 61.24.0.

**Pull request merged**